nav2_behavior_tree/plugins/action/follow_path_action.cpp

```diff
@@ -46,6 +47,22 @@ void FollowPathAction::on_wait_for_result()
     goal_.path = new_path;
     goal_updated_ = true;
   }
+
+  std::string new_controller_id;
+  getInput("controller_id", new_controller_id);
```

Comment:
I don't know how I feel about this one - I don't  think the controller / goal checkers should be able to be changed mid-execution. The goal checker I could almost make an argument for - but the planner ID is a bit more suspect. Can you explain the thinking here?

Also, why the `|=`? Those are the same as `=` for this situation (e.g. `goal_updated_` will always be `true` b/c the hardcoded-`true`)

Reply:
Hello. I have been out for some days.

Concerning the second question:

- Maybe I am missing something. From my understanding when goal_updated_ == true  the async_send_goal is executed. This change means that a new async_send_goal request is done if any of the inputs changes (goal or goal_checker or controller). It might be the case where the goal keeps the same but the goal checker changed, I think that information must be notified to the controller server. What I am missing?

First question: I more or less understand your concern. But IMHO, if we are working with topics we need this ability. Let me explain that:

**"The transactional configuration issue"**
Since controller_id, planner_id and goal_checker id inputs are topics if your application wants to use some "specific combination" you will face the issue of "the order". Input configuration topic messages may arrive unsorted and with some gap in time. 

The ideal behavior I think should be able to support the case of  "transactional one-shot configuration". Our solution is based on topics and if we do not allow to dynamically update planer, controller and/or goal checker we can face the situation where only one of them is updated and the FollowPathAction already started (what happens with the rest?)

From my understanding this issue may be "fair enough solved" using topics If we accept that controller, planners or goal checkers can be updated at any moment, even preempting the current execution.
The way I imagine the topics is used is keeping some kind of "continuous flow" of input messages showing the selected planner, controller and goal checker.

When I suggested the usage of services or actions in https://github.com/ros-planning/navigation.ros.org/issues/147 to mimic dynamic reconfigure.) I was thinking on this issue. I should have insisted about this.

What is your point of view this challenge? Is it something I am not seeing?

Reply:
My 2 cents, I generally think that it makes senses to expose the goal fields of an action in the BT action node calling it.
But to allow an update of these fields just feels weird to me (even for the `path`, though I understand why it is useful). After all, what a BtActionNode is doing underneath is to preempt the previous goal and executing a new one when `goal_updated_ == true`.
I feel that the same result could be achieved at the tree level with condition nodes and reactive fallbacks/sequences. 
(though to be logical, I should argue the same for `path`, propose to get rid completely of the `goal_updated_` logic and demonstrate how to do without it)

Reply:
The concept of preemption is usually annoying, I agree. But please, take into account the example I described. I think the concept Is very simple "making the navigation configurable from the ros network" (planner, controller and goal_update - specially the last two together)

Could you please provide a mock example of a xml bt-tree that did that? The only way I see if we do not accept preemption is introducing some BTAction that stores the updates of all the and update everything in one shot.

Reply:
I'm specifically asking 2 questions:
- Why are you using `|=` and not just `=`, it doesn't make any semantic difference to this application?
- If we should allow for changing controller ID / goal checkers mid-execution of the action.

---

Changing / selecting the planner / controller / algorithm makes alot of sense, so that when it gets updated **the next task** gets to use that algorithm. I did not intend this to be dynamically remapping in the middle of a running task. None of the logic of the controller server / action pipeline from the BT has been evaluated under those conditions. 

The planner doesn't really matter, you can remap every tick of the BT and that's OK since its a short running process that should be returning near immediately. The controller is unique that it is long-running. I'm not sure design-wise it would be smart to allow something like these to be changed in the middle of the navigation task run. Between preemption requests or goal cancellations, sure, because you're changing the nature of the request itself. You handle this in `on_tick()` that the new selected algorithms will be utilized for a new run. But changing the algorithm in the middle of its task run for the controller feels like such a fundamental change to the navigation request that doing so without a preemption/cancel feels off to me.

To be clear, if your argument is only that you want to enable this because you want to _guarantee_ that `new_controller` is used for the next task rather than `original_controller`, then maybe using services to get a response that the update was made is the most logical choice. If your argument though is that in the middle of a task from `A->B` you'd like to change the controller from `original_controller` to `new_controller` at any time, for any reason, anywhere, that's the "this feels wrong to me" starts to creep in. I don't think I'd ultimately block it if it was something you wanted to do, but I really want to have a discussion about this and make sure we all make smart choices. 

Not here nor there, but using dynamic reconfigure at run-time like that is a big no-no. Having done that in ROS1 wasn't a "solution" it was a hack. That was meant for local parameter tuning, not for use in steady-state and that was made pretty clear through its stability / documentation. 

---

@doisyg:  `goal` is special because that's how we communicate the action information to the behavior tree. `path` isn't anything special, its just a blackboard variable set as the output port of the planner and the input port of the controller BT nodes. There's no remapping / "changing" of these, that's just how the BT nodes talk to each other and a proper design pattern for blackboard/behavior tree communication. Even then, goal isn't actually hardcoded anymore, its all parameterizable and dynamic variably set in the XML. 

Also, @pabloinigoblasco wants to change the algorithm based on an external application. I 100% agree most people should use the BT contexts to set algorithms but he doesn't want to encode this information in the BT itself but from an external application (that's why we made the selector nodes).

Reply:
> - Why are you using |= and not just =, it doesn't make any semantic difference to this application?

You are right. My brain wires crossed here, I was thinking on the &=true case.

> - If we should allow for changing controller ID / goal checkers mid-execution of the action.

IMO, If the controller or the goal_checker change the nav_controller action server should be preempted. Are we aligned on this? 
Otherwise, if we have to wait for the FollowPath action to complete, even if these inputs change, there will be a race condition between the selected goal_checker and the selected controller. (you usually want to configure both together) 

> Changing / selecting the planner / controller / algorithm makes alot of sense, so that when it gets updated the next task gets to use that algorithm. 

Yes.

> I did not intend this to be dynamically remapping in the middle of a running task. None of the logic of the controller server / action pipeline from the BT has been evaluated under those conditions.

> The planner doesn't really matter, you can remap every tick of the BT and that's OK since its a short running process that should be returning near immediately. The controller is unique that it is long-running. I'm not sure design-wise it would be smart to change things like these in the middle of the navigation task run. Between preemption requests or goal cancellations, sure, because you're changing the nature of the request itself. 

> You handle this in on_tick() that the new selected algorithms will be utilized for a new run. But changing the algorithm in the middle of its task run for the controller feels like such a fundamental change to the navigation request that doing so without a preemption/cancel feels off to me.

Maybe I am not grasping something. From my perspective we are just preempting the nav_controller action server.

Do you mean here that we should cancel the action request and finish the FollowPath task if some of the inputs change? If so, that also sounds good to me.

> To be clear, if your argument is only that you want to enable this because you want to guarantee that new_controller is used for the next task rather than original_controller, then maybe using services to get a response that the update was made is the most logical choice. 

I think that topics can be enough to solve the most general case. It is true that we do not get any feedback if something failed on the selection though.

> If ​your argument though is that in the middle of a task from A->B you'd like to change the controller from original_controller to new_controller at any time, for any reason, anywhere, that's the "this feels wrong to me" starts to creep in. I don't think I'd ultimately block it if it was something you wanted to do, but I really want to have a discussion about this and make sure we all make smart choices.

From the point of view of the nav_controller I think it is equivalent to keep alive the same FollowPath action or finish that task and create a new one. The result is that the current action is preempted.
Aren't we just sending a new request to the action server if goal_updated_ == true? (preempting)

Nonetheless, if your concern is that the FollowPath action should finish when the goal_checker or the controller change, that is also good to me. Is that what you mean?

> Not here nor there, but using dynamic reconfigure at run-time like that is a big no-no. Having done that in ROS1 wasn't a "solution" it was a hack. That was meant for local parameter tuning, not for use in steady-state and that was made pretty clear through its stability / documentation.

I did not mean at all to use a dynamic_reconfigure solution, I just meant that ROS1 provided a solution that solved the issue of that potential "race condition" and we should find a solution that takes that into account.

Reply:
I think you know better how to define some kind of interruption with Behavior Trees. I have a few ideas but all of them are quite dirty.

The other alternative is to make a "buffer" BTNode with all the information to update and commit them in one shot. But that would make the BT more dirty and would show that the current selector approach is limited.

Reply:
> IMO, If the controller or the goal_checker change the nav_controller action server should be preempted. Are we aligned on this?

I don't understand the statement.

> From my perspective we are just preempting the nav_controller action server

Yes, which is what @doisyg and I are concerned about. We don't think that it necessarily makes sense to make it allowable to change that information mid-execution by dynamically set blackboard variables from the selector. The impression I was under was that the selector would be used to change the algorithms between preemptions, requests, cancellations, etc. I did not think about using the BT nodes to change it mid-execution without notifying anyone mid-request.

I'm more in favor of allowing for it than disallowing it, but it seems like a real potential for danger if the selector nodes are midused

Reply:
I understand your concerns but I think that safety issue you say already may happen now.

For example, If your goal checker only checks position and not speed, you may enter in another FollowPath task with a different controller that starts the navigation in some "unsafe" initial state.

Reply:
> I think that if you are really concerned about that safety issue it would be needed to introduce some safety checks (that you could optionally disabled) in the FollowPath action or in the nav_controller action server.

Or just not allow changing the algorithms mid-execution :wink:, but I'm leaning towards your side of things. @doisyg what do you think? Anything here convince you that this should be allowed? I'm leaning towards "yes" to allow it, even though I think its a bad idea to use it. I suppose its not really my place to tell people what they should/shouldn't do as long as it doesn't break general behavior.

Reply:
> Or just not allow changing the algorithms mid-execution wink

haha right! 

I am having a few more ideas (I do not want to spam though):

I see that one of the ways to finish the FollowPath is executing the cancel of the action request. What if the controllerSelector or the goalChecker or any other node sends cancel() to the nav_controller?

Some BT Pseudocode:
```
<ControllerSelector selected_controller="{selected_controller}>
<GoalCheckerSelector selected_goal_checker="{selected_goal_checker}">

<Parallel finish on any>
    <FollowPath controller_id="{controller}" goal_checker="selected_goal_checker"/>
    <InterruptFollowPath onchange="{selected_controller}"/>
    <InterruptFollowPath onchange="{selected_goal_checker}"/>
</Parallel>
```

Reply:
I think that could work for a custom BT of a user's design (probably not good for the default user BT) mostly just because its overkill for the average person -- but I suppose at that the selector nodes aren't even in the default BT (yet?). 

What is the goal you're trying to accomplish, specifically? Right now, `on_tick()` in only called on a new nav request (after a cancel) or a new Bt loaded. So that by itself would change the controller plugin algorihtms only when "new" without any additional work. It just wouldn't allow for preemption changes. I'm saying above that I'm OK with how it is now, as long as you are aware that your controller plugins need to be able to handle instantaneous changes if you use it in that way.

Reply:
> What is the goal you're trying to accomplish, specifically? 

The race condition: "changing in one controller and goal_checker". I need the FollowPath is restarted if the controller or the goal checker is updated.

> I'm saying above that I'm OK with how it is now, as long as you are aware that your controller plugins need to be able to handle instantaneous changes if you use it in that way.

Nonetheless, if the original proposal is fine, I am fine.

Reply:
I'm not sure I see a big issue with a race condition, what's the issue you're trying to avoid? I don't think cancel/request on a change of controller ID / goal checker is necessary. I think if I were designing an industrial / consumer robot navigation system, I'd want the opposite: for a cancel/request, I'd want the newest controller ID / goal checker. See what I mean?

Reply:
> I'm not sure I see a big issue with a race condition, what's the issue you're trying to avoid?

This issue:
![image](https://user-images.githubusercontent.com/13334595/113069502-cc2a3a80-91c0-11eb-9a0a-9ea3d7e741af.png)


The previous case would be an incorrect execution, because the decision making about "the mode of navigation" should be executed in one shoot.

> I'd want the opposite: for a cancel/request, I'd want the newest controller ID / goal checker. See what I mean?

I agree with this sentence. But I think that the issue described in the previous drawing is still there.
 
> This feature as is works when it works, the problem is that many people might not understand the stability of swapping controllers on the fly at full speed while in an existing navigation request.

Indeed, any controller switching may be dangerous. With or without the controller selector.

Reply:
Well if you publish the selector topics before the action request, you're probably in good hands that things will work out but this is the nature of the beast of doing things this way vs in subtree contexts like we suggest. I can't really offer much help as to what to do here at that point because it starts to dig deeply into your specific systems.

Reply:
>  This is also why we allow folks to specify their own BT XMLs versus only the default ones we can design for the typical user

This was the discussion for months, I see your point of view totally legit. But I think it is also legit to use "the navigation stack as a remote service to request motions tasks". This approach also totally fits the philosophy of work of a distributed framework as ROS2.

Discussions aside, lets try to go forward with a solution that make everybody happy.  

I would like to suggest a couple more of ideas that can make the selectors better and may remove the need of dynamic blackboard variables and solve some of the problems we are discussing. I am not just thinking in my case, but in a general and flexible usage of the selectors.

If they are not accepted it is okay, we can keep the current basic implementation (I can even remove that preemption of the nav_controller if feel better with that). Nonetheless I would like to know your opinion about them:

**Service inputs**
After the discussion I now think the input of the selectors should move from topic to service:

 - That may be good for a better the synchronization in the client that needs to configure different parameters. The client can verify the configuration was properly received before sending the final goal request.
 - Service result could be also be used to provide some feedback about the success/failure of the change (maybe checking if the controller/planner/goal_checker exists? maybe rejecting the change if the nav_controller is busy?) 
   
**Blocking selectors**
I would also introduce in the selectors a blocking attribute. With this approach you could define optional or mandatory configurations that must be specified mandatory for each NavigateToGoal request.  
That would be also good for the case you specify a new bt file in the NavigateToGoal message avoiding race conditions.
 
Something like this:
  
```xml
<btree>
<sequence>
<!-- configuration stage -->
<parallel> 
    <PlannerSelector blocking="true" .../>
    <ControllerSelector blocking="true" .../>
    <GoalChecker blocking="true" .../>
</parallel>
<-- navigation stage-->
<ComputePlan ...>
<FollowPath .../>
</sequence>
</btree>
```

Reply:
>  lets try to go forward with a solution that make everybody happy

I thought I already seeded this point to you that we could keep the dynamically set controller ID / goal checkers? My only comment was that its on the user to make sure that your controllers have some stability to operate in that fashion. As long as the controller makes use of https://github.com/ros-planning/navigation2/blob/d8bd3c5ea8d196ad7f0bace1edaeaa8e55f06944/nav2_controller/src/nav2_controller.cpp#L386 the inputted current velocity, I think things should be OK for the most part. I know for instance RPP will, not I can't speak to the others.

> Service inputs

Services make a ton of sense on a first glance. This is exactly what they do well. I did not explicitly suggest that because of the spinning issue to process requests. We'd either have to create a new node on the network to handle those which would be alot of overhead or a separate single threaded executor spinning a callback group (like the service client in utils was changed to recently). The second is the subject of https://github.com/ros-planning/navigation2/issues/2251 which I'm still not 100% certain if creating the new executor adds a new DDS participant to the network (which is ~75% the overhead of just a new node). If I knew for certain one way or another, I could discount or immediately embrace the service idea. 

> Preload behavior tree service

That's a good point on race conditions if its even alive, I didn't think about using selectors in the case of also changing the BT in the `NavigateToPose.action`. One way you could actually help resolve that is to have your publisher of the selector **and** the subscriber be transient-local. that way you get the "latched" behavior of ROS1 so that if a message is published **after** the subscriber is created, the message still makes it there. I think that would largely resolve the concern that this suggestion is fixing. I'd actually suggest adding transient local QoS to the subscribers in all 3 of these right now from that motivating example. 

> Blocking selectors

How different would that be from specifying non-sensical default values in the selector node? If they didn't make sense, then when one was requested, then it would fail, and fail the request the same way something blocking would also fail the request. Though semantically that seems dubious. A boolean for `blocking` that is default to `false` would be acceptable. Alternatively, we could also make the default algorithm be an empty string. And when its an empty string, we enable `blocking` behavior because clearly that is not a valid algorithm to select. That would implicitly do blocking when no default is acceptable without adding a new parameter. It would only require a minor update to the code and an update to documentation mentioning this is the fact for the default plugins ports.

Tl;dr
- I like services too, but I'm not sure about #2251, if you knew for a fact (or could confidently determine it) that it didn't add a new DDS participant due to the new context, then that's the best solution
- QoS for all 3 to `rclcpp::KeepLast(1)).transient_local().reliable()`, and your publisher being transient local would be good to make sure even in race conditions of a new BT that it was received 
- Blocking is a good feature idea! Though I think we can do it if the `default` port is `""`, then block / fail. Does blocking implicitly when there is no valid default value which I think is your intent without adding a new port.

Reply:
I read carefully everything you described. Thanks for the detailed answer.

> - I like services too, but I'm not sure about #2251, if you knew for a fact (or could confidently determine it) that it didn't add a new DDS participant due to the new context, then that's the best solution

Okay. I understand that this must wait, but lets keep an eye to this since it looks to be the best solution.

> - QoS for all 3 to rclcpp::KeepLast(1)).transient_local().reliable(), and your publisher being transient local would be good to make sure even in race conditions of a new BT that it was received

Okay, this is something I am about to commit and push. However, I think this is not a solution for the race condition, maybe it is for the first navigation request, but I think it is not a solution second motion request. (see pseudocode below)

> - Blocking is a good feature idea! Though I think we can do it if the default port is "", then block / fail. Does blocking implicitly when there is no valid default value which I think is your intent without adding a new port is "", then block / fail. Does blocking implicitly when there is no valid default value which I think is your intent without adding a new port.

I think this need a bit more discussion because I am not sure if we are thinking in the same solution. 
What is blocking for you? 
- Rejecting the navigateToGoal request? 
- With "blocking" I mean "wait until you receive a new input". That is: replacing the navigation request from "one shot action" to some "interactive protocol between the user and the action server". I think this "weird" concept make the point of view different.

My main concern is still the race condition. I think that "reject the request" only would solve the issue for the first time.
Let me show the example I have in mind "two consecutive navigation requests":

```
- select planner_A
- select controller_A
- select goal_checker_A
[Here is the race condition, the request returns fail if elements were not properly selected,
Lets assume this first attempt worked fine]
- navigate to goal X1
- wait to finished navigation
[Navigation returns success]

[Now second motion request]
- select planner_B
- select controller_B
- select goal_checker_B
[Race condition, select_controller_B was late and controller_A was still selected]
- navigate to goal X2
- wait to finished navigation
[Navigation returns success]
```

The second NavigateToGoalPose request would work if select_controller was blocking. On the other hand, if we just return fail if the "controller is not known" would end up in some incorrect navigation execution of the second motion.

The "workaround" I am having in mind is just using `Wait(seconds)` between the selections and each navigate request. But that is ugly and blind (the client never knows if things are all right). Services would have been enough (I know they are frozen)

I think this race condition is not only something of my case. I think it is the nature of the solution we are providing for selectors and may affect to anyone using them.

Reply:
On blocking: what I suggest is:
- Make the default planner / controller / goal checker IDs an empty string
- If the current is empty string in the `tick` because non set and default was an empty string, then nothing has been set, fail the bt status -- blocking motion
- If the current is non-empty string in `tick`, then something was set and no need to block
- If the default was set to a non-empty string in the XML, that disables blocking because there is a logical default

So without a specific `blocking` parameter, we can get that behavior by just checking the current algorithm ID and the default (if set to empty to enable -- meaning semantically that there IS no default and there MUST be a value set to pass). 

---

If you're concerned about bleeding between runs, you could specify a BT to use in the action request so that a new tree is loaded - though that only works if you're constantly changing trees between requests. I suppose we could have a `force_reload` parameter / action request field so that the BT is reloaded every time a navigation request comes in to prevent that.

Another options with the selector nodes is that you can override the `halt` function to reset the current back to empty string (if the default is empty string to enable blocking) when the BT exists so that each run will be a clean state execution. That's probably the cleanest solution to this particular problem, but the `force_reload` is the most general to solve all such problems in the future (though very inefficient and adds latency)